### PR TITLE
Decouple time.Time parsing from empty interface behavior.

### PR DESCRIPTION
**Pull request by benluddy**
#### PR Was Proposed and Welcomed in Currently Open Issue

- [x] This PR was proposed and welcomed by maintainer(s) in issue #501
- [x] Closes or Updates Issue #501

#### Checklist (for code PR only, ignore for docs PR)

- [x] Include unit tests that cover the new code
- [x] Pass all unit tests 
- [x] Pass all 18 ci linters (golint, gosec, staticcheck, etc.)
- [x] Sign each commit with your real name and email.  
      Last line of each commit message should be in this format:  
      Signed-off-by: Firstname Lastname <firstname.lastname@example.com>
- [x] Certify the Developer's Certificate of Origin 1.1
      (see next section).

#### Certify the Developer's Certificate of Origin 1.1

- [x] By marking this item as completed, I certify 
      the Developer Certificate of Origin 1.1.

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
660 York Street, Suite 102,
San Francisco, CA 94110 USA

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

### CODE REVIEW

File: decode_test.go [hunk 1]
```diff
@@ -3759,7 +3773,7 @@ func TestDecodeInvalidTagTime(t *testing.T) {
 			name:          "Tag 1 with negative integer overflow",
 			data:          hexDecode("c13bffffffffffffffff"),
 			decodeToTypes: []reflect.Type{typeIntf, typeTime},
-			wantErrorMsg:  "cbor: cannot unmarshal tag into Go value of type time.Time",
+			wantErrorMsg:  "cbor: cannot unmarshal negative integer into Go value of type time.Time (-18446744073709551616 overflows Go's int64)",
```
Comment (benluddy) [inline decode_test.go:3776]: My understanding is that overflow was causing parse to return `math.Big` in these cases, so the previous error came from the default case of the type switch in parseToTime.

File: decode_test.go [hunk 2]
```diff
@@ -3703,11 +3704,24 @@ func TestDecodeTimeError(t *testing.T) {
...
 			wantErrorMsg: "cbor: cannot unmarshal negative integer into Go value of type time.Time",
 		},
+		{
+			name: "untagged byte string content cannot be decoded into time.Time with DefaultByteStringType string",
+			opts: DecOptions{
+				TimeTag:               DecTagOptional,
+				DefaultByteStringType: reflect.TypeOf(""),
+			},
+			data:         hexDecode("54323031332d30332d32315432303a30343a30305a"),
+			wantErrorMsg: "cbor: cannot unmarshal byte string into Go value of type time.Time",
```
Comment (benluddy) [inline decode_test.go:3714]: This reproduces #501 (on master, a nil error is returned instead).

File: decode.go [hunk 3]
```diff
@@ -1254,49 +1252,79 @@ func (d *decoder) parseToTime() (tm time.Time, err error) {
...
+		if val > math.MaxInt64 {
+			if val == math.MaxUint64 {
+				// Maximum absolute value representable by negative integer is 2^64,
+				// not 2^64-1, so it overflows uint64.
+				return time.Time{}, &UnmarshalTypeError{
+					CBORType: t.String(),
+					GoType:   typeTime.String(),
+					errorMsg: fmt.Sprintf("-18446744073709551616 overflows Go's int64"),
+				}
+			}
```
Comment (fxamacker) [inline decode.go:1298]: Thanks for preventing overflow :+1:

File: decode.go [hunk 4]
```diff
@@ -1254,49 +1252,79 @@ func (d *decoder) parseToTime() (tm time.Time, err error) {
...
+			return time.Time{}, &UnmarshalTypeError{
+				CBORType: t.String(),
+				GoType:   typeTime.String(),
+				errorMsg: fmt.Sprintf("-%d overflows Go's int64", val+1),
+			}
+		}
+		return time.Unix(int64(-1)^int64(val), 0), nil
+	case cborTypePrimitives:
+		_, ai, val := d.getHead()
```
Comment (fxamacker) [inline decode.go:1307]: We also need to handle CBOR nil and undefined here by returning `time.Time{}` and no error.

File: decode.go [hunk 5]
```diff
@@ -1254,49 +1252,79 @@ func (d *decoder) parseToTime() (tm time.Time, err error) {
...
+	case cborTypeNegativeInt:
+		_, _, val := d.getHead()
+		if val > math.MaxInt64 {
+			if val == math.MaxUint64 {
+				// Maximum absolute value representable by negative integer is 2^64,
+				// not 2^64-1, so it overflows uint64.
+				return time.Time{}, &UnmarshalTypeError{
+					CBORType: t.String(),
+					GoType:   typeTime.String(),
+					errorMsg: fmt.Sprintf("-18446744073709551616 overflows Go's int64"),
```
Comment (fxamacker) [inline decode.go:1296]: nit: `fmt.Sprintf` is not needed here

**Review by fxamacker — COMMENTED** [review]
Thanks for opening this PR!  Looks great!

I left a comment about the need to handle decoding CBOR nil and CBOR undefined to `time.Time{}`.

File: decode.go (hunk 4 above)
Comment (benluddy) [reply decode.go:1307]: As far as I can tell, that case only occurs when unrecognized tags in the input are combined with TimeTag=DecTagIgnored. I made a table of the current behavior on master in https://gist.github.com/benluddy/0f43e72070c129363e5faa6afb539921, and some of the DecTagIgnored behaviors look a little asymmetric to me:

1. `0(null)` and `1(null)` are rejected by the built-in tag content type validation, but `3735928559(0(null))` is not.
2. `3735928559(null)` decodes to Time's zero value, but `null` is a no-op.
3. `1(3735928559(null))` is rejected by the built-in tag content type validation, but `3735928559(1(3735928559(null)))` decodes to Time's zero value.

I'm not sure how to reconcile these. Maybe the simplest thing to do is bypass the built-in tag content type validation for tags 0 and 1 when TimeTag is DecTagIgnored, and accept that swallowing tags results in some permissive behavior?

Comment (fxamacker) [reply decode.go:1307]: Thanks for creating the table of all input combinations!  :heart_eyes:

`DecTagIgnored` option was meant to handle unknown tags.

I agree that `DecTagIgnored` behaviors have some inconsistency.

Some ideas for consideration with `TimeTag=DecTagIgnored`:
- decoding a tagged `null` should be a no-op to be consistent with `null`.
- supported known tags should be validated to be consistent with `0(null)`, etc.

For example:
- `3735928559(0(null))` and `3735928559(1(3735928559(null)))` should be rejected by tag validation.
- `3735928559(null)` should be a no-op as `null`.  

Thoughts?

Thanks again for your insights and contributions!

Comment (benluddy) [reply decode.go:1307]: > - `3735928559(0(null))` and `3735928559(1(3735928559(null)))` should be rejected by tag validation.

Although I'm not personally using `DecTagIgnored`, that would be my preference given both examples are well-formed but invalid. There is precedent in other options for decoding potentially-invalid input, like `UTF8DecodeInvalid`, but it's easy to imagine a use case for that one where the input has already been validated or will be validated after decode. After decoding `3735928559(0(null))` with `DecTagIgnored`, a caller would have no signal to detect that there had been an invalid input.

> - `3735928559(null)` should be a no-op as `null`.

This also makes sense to me. The existing behavior for 1-tagged floats with NaN or infinite values (as briefly referenced in https://www.rfc-editor.org/rfc/rfc8949.html#section-3.4.2-6) to support use cases that want to explicitly represent a present-and-zero-valued time. I don't think that is documented -- I'll include documentation updates for that as well as for `TimeTag`.

**Comment by benluddy** [comment]
@fxamacker Thanks for the discussion in https://github.com/fxamacker/cbor/pull/503#discussion_r1513768915! I've added tests for those edge cases and resolved them according to your suggested behaviors. PTAL!

File: decode.go [hunk 6]
```diff
@@ -1239,64 +1262,97 @@ func (d *decoder) parseToTag(v reflect.Value) error {
...
 		if d.dm.timeTag == DecTagIgnored {
-			// Skip tag number
+			// Skip all enclosing tags
 			for t == cborTypeTag {
 				d.getHead()
 				t = d.nextCBORType()
 			}
+			if d.nextCBORNil() {
+				return time.Time{}, false, nil
+			}
```
Comment (fxamacker) [inline decode.go:1278]: We need to skip CBOR nil here before returning. 

```suggestion
			if d.nextCBORNil() {
			        d.skip() // skip CBOR nil
				return time.Time{}, false, nil
			}
```

File: decode.go [hunk 7]
```diff
@@ -87,7 +87,8 @@ import (
 // To unmarshal a CBOR text string into a time.Time value, Unmarshal parses text
 // string formatted in RFC3339.  To unmarshal a CBOR integer/float into a
 // time.Time value, Unmarshal creates an unix time with integer/float as seconds
-// and fractional seconds since January 1, 1970 UTC.
+// and fractional seconds since January 1, 1970 UTC. As a special case, Infinite
+// and NaN float values decode to time.Time's zero value.
```
Comment (fxamacker) [inline decode.go:91]: Thank you for adding the comment about special case! :+1:

File: decode.go [hunk 8]
```diff
@@ -476,8 +477,28 @@ type DecOptions struct {
...
+	//   time.Time does not modify the destination value.
+	//
+	//   DecTagOptional: Untagged text strings are decoded as though
+	//   enclosed in tag 0. Untagged unsigned integers, negative integers,
+	//   and floating-point numbers are decoded as though enclosed in tag
+	//   1. Tags other than 0 and 1 will produce an error on attempts to
+	//   decode them into a time.Time.
+	//
+	//   DecTagRequired: Only tags 0 and 1 can be decoded to time.Time. Any
+	//   other input will produce an error.
```
Comment (fxamacker) [inline decode.go:501]: Thanks so much for improving the documentation! :+1:

**Review by fxamacker — COMMENTED** [review]
Thanks for updating this PR and adding detailed documentation!  

I left a comment related to skipping data before returning from `parseToTime()`.

**Review by fxamacker — APPROVED** [review]
Looks great!  Thanks so much for contributing this PR!  :+1:
